File: 프로그래머스/lv1/17682. ［1차］ 다트 게임/［1차］ 다트 게임.java
class Solution {
    public int solution(String dartResult) {
        int answer = 0, tempNum = 0, idx = 0;
        int[] dart = new int[3];
        String numStr = ""; // 정수를 문자열로 합쳐놓고 문자가 나오면 정수로 변환. 10이상의 수 대비
        for (int i = 0; i < dartResult.length(); i++) {
            char c = dartResult.charAt(i);
            if (c >= '0' && c <= '9') {
                numStr += String.valueOf(c);
            }
            else if (c == 'D' || c == 'S' || c == 'T') {
                tempNum = Integer.parseInt(numStr);
                if (c == 'D') {
                    dart[idx++] = (int)Math.pow(tempNum, 2);
                }
                else if (c == 'S') {
                    dart[idx++] = (int)Math.pow(tempNum, 1);
                }
                else if (c == 'T') {
                    dart[idx++] = (int)Math.pow(tempNum, 3);
                }
                numStr = "";
            }
            else {
                if (c == '*') {
                    if (idx >= 2) {
                        dart[idx-1] *= 2;
                        dart[idx-2] *= 2;
                    }
                    else {
                        dart[idx-1] *= 2;
                    }
                }
                else if (c == '#') {
                    dart[idx-1] *= (-1);
                }
            }
            
        }
        answer = dart[0] + dart[1] + dart[2];
        return answer;
    }
}

File: 프로그래머스/1/17682. ［1차］ 다트 게임/［1차］ 다트 게임.py
def solution(dartResult):
    answer = 0
    tempNum = ""
    dart = [0]*3
    idx = 0
    for i in range(len(dartResult)):
        # 문자가 들어오기 전까지 숫자는 일단 합쳐둔다. 두자리수 대비하기 위함

        if dartResult[i] == 'S':
            dart[idx] += int(tempNum)
            idx += 1
            tempNum = ""
        elif dartResult[i] == 'D':
            dart[idx] += int(tempNum) ** 2
            idx += 1
            tempNum = ""
        elif dartResult[i] == 'T':
            dart[idx] += int(tempNum) ** 3
            idx += 1
            tempNum = ""
        elif dartResult[i] == '*':
            dart[idx-1] *= 2
            if idx > 1:
                dart[idx-2] *= 2
        elif dartResult[i] == '#':
            dart[idx-1] *= -1
        
        else: 
            tempNum += dartResult[i]
    
    return sum(dart)

'''
문자 나오면 그 앞에 값에 제곱, 세제곱 등을 해서 합선처리

*나오면 그전에 합산된 값에 x2
#나오면 그전 합산된 값에 -1 곱함
'''